Add possibility to stop execution after specific number of errors

piodul opened the pull request:
This PR adds support for the `-error-limit` flag from the original scylla-bench. The flag allows to limit the total number of errors that can occur during a run before the run is actually stopped.

cvybhu commented on src/run.rs:
```diff
@@ -314,6 +324,7 @@ mod tests {
             rate_limit_per_second: None,
             operation_factory: Arc::new(FnOperationFactory(f)),
             max_retries_per_op: 0,
+            max_retries_in_total: usize::MAX,
```
By default `retiries_in_total` are unlimited, but `retries_per_op` are set to 0. This is a bit inconsistent.

piodul replied:
I didn't use `max_retries_in_total` here because I didn't want to change the existing tests that used `max_retries_per_op`, however I don't think it's very bad so ultimately I changed `max_retries_in_total` here to 0 and adjusted the affected tests.

piodul commented on src/run.rs:
```diff
@@ -135,7 +139,13 @@ impl WorkerSession {
             Err(err) if self.context.should_stop() => Err(err),
-            Err(_) => {
+            Err(err) => {
+                let countdown = self.context.retry_countdown.fetch_sub(1, Ordering::Relaxed);
```
Yes, it will just wrap around. This behavior is defined in the documentation: https://doc.rust-lang.org/std/sync/atomic/struct.AtomicU64.html#method.fetch_sub

piodul commented on src/run.rs:
```diff
@@ -135,7 +139,13 @@ impl WorkerSession {
+                if countdown == 0 {
+                    // We have exhausted our global number of allowed retries.
+                    self.context.ask_to_stop();
+                    return Err(err);
```
Hmm, right, as far as I know the original scylla-bench goes a step further and even prints all errors that happened during the test... I'll look closer into how error reporting is done there.

piodul replied:
After taking a closer look - nope, they just return the last error which failed due to the global limit + an error for each task which failed due to subsequent failures on a single task (`error-at-row-limit`). I'll adjust how we print errors accordingly.

piodul commented on src/bin/cql-stress-scylla-bench/main.rs:
```diff
@@ -146,7 +146,7 @@ async fn prepare(args: Arc<ScyllaBenchArgs>, stats: Arc<ShardedStats>) -> Result
         operation_factory,
         max_retries_per_op: args.max_retries_per_op as usize,
-        max_retries_in_total: usize::MAX,
+        max_retries_in_total: args.max_retries_in_total as usize,
```
Hmm, I used `usize` here because `max_retries_per_op` was also of this type. I'll see if changing the type of `max_retries_per_op` to `u64` makes sense.

piodul replied:
OK, done. There are indeed less casts now.

piodul commented on src/bin/cql-stress-scylla-bench/args.rs:
```diff
@@ -429,6 +429,16 @@ impl ScyllaBenchArgs {
+        if self.max_retries_in_total == u64::MAX {
+            println!("Max error number:\t unlimited");
+        } else {
+            println!("Max error number:\t {}", self.max_retries_in_total);
```
Good point, fixed

piodul commented on src/bin/cql-stress-scylla-bench/args.rs:
```diff
@@ -429,6 +429,16 @@ impl ScyllaBenchArgs {
+        if self.max_retries_per_op == u64::MAX {
+            println!("Max error number at row: unlimited");
+        } else {
+            println!("Max error number at row: {}", self.max_retries_per_op);
```
Yes. Fixed.

piodul commented on src/run.rs:
```diff
@@ -60,6 +62,8 @@ impl WorkerContext {
         Self {
             operation_counter: AtomicU64::new(0),
 
+            retry_countdown: AtomicU64::new(config.max_retries_in_total as u64),
```
Changed

piodul commented:
v2:
- Fixed review comments
- The s-b frontend now prints all errors returned from tasks instead of only the first one
- Slightly changed the run task's behavior in case the run stops in the middle of the retry loop

I rebased the PR on top of https://github.com/piodul/cql-stress/pull/37 for the CI to pass, so please review that one first. As for the commit that changes the behavior on error (the last one), I found it when improving the error printing logic and included it here instead of sending separately in order to reduce a chance of merge conflicts.

cvybhu commented on src/run.rs:
```diff
@@ -147,11 +149,19 @@ impl WorkerSession {
             }
-            Err(err) if self.trial_idx >= self.context.max_retries_per_op => Err(err),
+            Err(err) if self.trial_idx >= self.context.max_retries_per_op => {
+                Err(err.context(format!(
+                    "Maximum number of retries per operation exceeded ({})",
+                    self.context.max_retries_per_op,
+                )))
+            }
```
The argument that user passed talked about the maximum number of errors, so maybe it would be better to mention errors in the message as well.

piodul replied:
I didn't want to do it because the terminology used in the context talks about retries, not errors - however the error counts might make more sense to users, so I'll use that.

piodul commented on src/bin/cql-stress-scylla-bench/stats.rs:
```diff
@@ -274,7 +279,16 @@ impl StatsPrinter {
+                out,
+                "\nFollowing critical errors were caught during the run:"
+            )?;
+            for err in errors {
+                // The {:#} syntax makes sure that the error is printed in one line
+                writeln!(out, "    {:#}", err)?;
```
ahyhow's documentation suggests using `{:#}` here: https://docs.rs/anyhow/latest/anyhow/struct.Error.html#display-representations

piodul commented on src/bin/cql-stress-scylla-bench/stats.rs:
```diff
@@ -252,7 +252,12 @@ impl StatsPrinter {
+        stats: &Stats,
+        errors: &[anyhow::Error],
+        out: &mut impl Write,
+    ) -> Result<()> {
```
There will be at most as many error as the `concurrency` parameter. So, in the worst case, it will be thousands.

Besides, this is what scylla-bench does, and I'm striving for compatibility - maybe not with regards to exact error strings, but formatting at least...

piodul commented on src/run.rs:
```diff
@@ -260,17 +273,19 @@ async fn do_run(config: Configuration, stop_receiver: oneshot::Receiver<()>) ->
-    // TODO: Collect all errors and report them
     while let Some(worker_result) = worker_handles.next().await {
         if let Err(err) = worker_result {
-            result = Err(err);
-            ctx.ask_to_stop();
+            errors.push(err);
         }
     }
```
Right, it doesn't make sense. Good catch. ask_to_stop is now only called when the global number of retries is exhausted, but not when the local number of retries is exhausted. I'll fix it and add a test.

piodul replied:
Done, I also added a regression test which checks that an error from one task stops others.

piodul commented on src/bin/cql-stress-scylla-bench/gocompat/strconv.rs:
```diff
@@ -28,7 +28,7 @@ pub fn parse_int(mut s: &str, bit_size: u32) -> Result<i64> {
     // Detect the sign
     let negative = s.starts_with('-');
-    s = s.strip_prefix(&['-', '+']).unwrap_or(s);
+    s = s.strip_prefix(['-', '+']).unwrap_or(s);
```
Done, but clippy started to complain again. I will rebase again after https://github.com/piodul/cql-stress/pull/38 is merged, as it contains the necessary fixes.

piodul commented:
v3:

- Rebased (will need to do it once again after https://github.com/piodul/cql-stress/pull/38)
- Fixed a bug that prevented the benchmark from stopping when one task fails with an error, added a test for it
- Added a bit more explanation in the comment for `decrement_global_retry_counter`

piodul commented:
Sorry for the long delay, I completely forgot about this PR...

piodul commented:
It looks like I slightly misunderstood the meaning of the original `max-errors-at-row` flag. In original `scylla-bench`, it does not cause the operation to be retried for the same piece of data, instead it will be skipped. I'll update the PR according to that semantics.

There are more flags in the original scylla-bench that configure proper retries, but they will be implemented in a separate PR.

piodul commented:
v4:
- Aligned the semantics of the `error-at-row-limit` flag with what happens in scylla-bench
- Rebased on top of master to get rid of clippy complaints.

piodul commented:
@cvybhu review ping

cvybhu commented on src/run.rs:
```diff
@@ -497,4 +524,73 @@ mod tests {
+            let mut cfg = make_test_cfg(move || Op::new(shared_counter.clone()));
+            cfg.concurrency = error_count;
+            cfg.max_consecutive_errors_per_op = 1; // We need to allow the runner to retry individual failures
+            cfg.max_errors_in_total = retry_limit;
```
I think it would be good to rename `retry_limit` to `max_errors_allowed`, it would be easier to understand.

cvybhu commented on src/bin/cql-stress-scylla-bench/args.rs:
```diff
@@ -429,6 +429,22 @@ impl ScyllaBenchArgs {
+        } else {
+            println!(
+                "Max error number at row: {}",
+                self.max_consecutive_errors_per_op as u128 + 1,
```
This `+1` could be removed after removing `sub(1)`

cvybhu replied:
Also, there's no `\t` in this particular print, is this intentional?

cvybhu commented on src/run.rs:
```diff
@@ -149,11 +151,17 @@ impl WorkerSession {
-                Err(err)
+                Err(err.context(format!(
+                    "Maximum number of errors allowed per operation exceeded ({})",
+                    self.context.max_consecutive_errors_per_op as u128 + 1,
```
And this `+1` as well

cvybhu commented on src/bin/cql-stress-scylla-bench/args.rs:
```diff
@@ -330,7 +330,7 @@ where
         // Zero means unlimited tries,
         // and #tries == #retries + 1,
         // therefore just subtract with wraparound and treat u64::MAX as infinity
-        let max_retries_per_op = max_errors_at_row.get().wrapping_sub(1);
+        let max_consecutive_errors_per_op = max_errors_at_row.get().wrapping_sub(1);
```
I don't understand what the comment about retries is trying to say, previously it made some sense, as the variable used to specifiy the number of retries, but now it would be good to update it so that it explains the new variable.

cvybhu replied:
The test `test_retrying()` sets `max_consecutive_errors_per_op = 0`, and then runs an operation which fails the first time it's executed. It's expected that the run will fail.

With `sub(1)` this would correspond to `error-at-row-limit = 1`. That would mean that `error-at-row-limit = 1` doesn't allow any errors, this sounds wrong.

cvybhu reviewed (changes requested):
Reviewed, sorry it took so long.

cvybhu commented:
My current understanding of these flags is:

`error-at-row-limit` - The maximum number of consecutive errors allowed. After exceeding it, the workflow will terminate with an error.

For example, with `error-at-row-limit = 3`:

This is ok (3 consecutive errors):
```
intial query - ERROR
retry 1 - ERROR
retry 2 - ERROR
retry 3 - OK
```

This is not ok (4 consecutive errors):
```
intial query - ERROR
retry 1 - ERROR
retry 2 - ERROR
retry 3 - ERROR
> The whole run fails <
```


`error-limit` - The number of errors after which the run should fail. Having `error-limit - 1` errors is fine, but one more and the run fails.
For example, with `error-limit = 3`, `max_errors_in_total = 2` (`retry_countdown` is intialized to `2`):

This is ok:
```
initial query - ERROR (`retry_countdown.fetch_sub(1) = 2`)
retry 1 - ERROR (`retry_countdown.fetch_sub(1) = 1`)
retry 2 - OK
```

This is not ok:
```
initial query - ERROR (`retry_countdown.fetch_sub(1) = 2`)
retry 1 - ERROR (`retry_countdown.fetch_sub(1) = 1`)
retry 2 - ERROR (`retry_countdown.fetch_sub(1) = 0`)
> The whole run fails <
```

muzarski commented on src/bin/cql-stress-scylla-bench/args.rs:
```diff
@@ -332,6 +339,9 @@ where
         let max_consecutive_errors_per_op = max_errors_at_row.get().wrapping_sub(1);
 
+        // Similar to above
+        let max_errors_in_total = max_errors.get().wrapping_sub(1);
```
Just to make sure i understand correctly - this option (unlike `error-at-row-limit`) defines the number of errors after which the whole run fails. In other words, when `error-limit = 1`, we immediately fail the run after the first error occurs. Am i right?

If so, then this `wrapping_sub(1)` looks semantically correct to me (looking at the `WorkerContext::decrement_global_retry_counter`).

muzarski reviewed (commented):
There is also a matter of rebasing on top of master, and adjusting the `cassandra-stress` frontend to these changes.

After analyzing the code of `cassandra-stress`:
- the `max_consecutive_errors_per_op` config option should be by default set to 9 (as it is now). It's defined via `-errors retries` (will be introduced soon) parameter. This means that we should try to perform each operation `1 + retries` number of times.
- there is no corresponding option for `max_errors_in_total` in c-s. It can be set to `u64::MAX`.

roydahan commented:
@muzarski (cc @fruch ) this is one of the items must be completed before we can switch SCT cases to use cql-stress.

roydahan commented:
Actually I see it's specific for s-b, which is second phase.
